README.md
# RGB-Pi

Download at rgb-pi.com

------------------------------------------------------------------------------

To use the cable in Retropie, Raspbian or any other system you can insert this lines in config.txt

Warning, this mode only support single resolution, We only support multiple resolutions in RGB-Pi OS

    dtparam=audio=on
    dtoverlay=pwm-2chan,pin=18,func=2,pin2=19,func2=2
    dtoverlay=rgb-pi
    enable_dpi_lcd=1
    display_default_lcd=1
    dpi_output_format=6
    dpi_group=2
    dpi_mode=87
    hdmi_timings=320 1 16 30 34 240 1 2 3 22 0 0 0 60 0 6400000 1

And add the file /overlays/rgb-pi.dtbo from the repo .dtbo if your OS is 4.1 or .dtb for 4.0


changelog.md
**RGB-Pi OS 2 (Alpha - 20200501) Soon**
```
- Added default OS turbo mode for improved performance.
- Changed to native video driver for improved frame latency.
- Changed to latest retroarch + core updates.
- Added experimental Sharp x68000.
- Changed FB Alpha by FB Neo.
- Added NEOGEO CD system.
- Changed system menu and system options interface.
- Added full CW and CCW UI rotation modes. 
- Added new Skin UI engine.
- Added new unified search + scrap engine.
- Added support for video previews.
- Changed game list and favorite data are now saved per unit.
- Added new game list options menu.
- Added quick sort modes by letter, year, players and system.
- Added forward and backward for all the quick search modes.
- Added Virtual Keyboard for data entry.
- Added Kiosk mode password protected. 
- Added 5 new languages (EN,ES,FR,DE,IT,PT,JP).
- Added display menu music title.
- Added rom folder auto-regeneration.
- Added NFS2/3 and NFS4 modes.
- Added NFS check to avoid overwrite hi-scores.
- Added current selected external unit icon.
- Added three new screen savers, black screen, video previews and CRT power off.
- Added Image List menu type.
- Added system power and temperature alarms.
- Added arcade FreePlay mode.
- Added controller and button layout association. 
- Fixed some firewall blocking issues.
- Fixed SSID and Password char issues.
- Fixed all remaps on real RGB-Pi JAMMA mode.
- Changed unified Start hotkey for solving issues with PC-Engine, MegaDrive and JAMMA.
- Changed main music to Evan King chiptune albums.
- Added new custom music album option.
- Added new Bluetooth menu.
- Added new Netplay menu.
- Added new External Unit menu.
- Changed all RetroArch options to specific menus.
- Changed EQ menu.
- Changed controller mapping by direct selection.
- Added new manual power off option.
- Added power off sequence image indicators.
- Removed auto scan games on boot for super fast load.
- Added videoplayer UI integration.
- Removed cable check restriction.
```
**RGB-Pi OS (Beta - Update) 2019 08 06**
```
- Fixed UI font. Now is fully monospace
- Fixed cls error cls on system boot
- Fixed some english translations
- Fixed performance issue in arcade image selector screen
- Fixed advmame paths for reading from the current selected unit
- Fixed vertical games overlay displayed on horizontal games
- Added apostrophe character support in Wifi configuration
- Added .hdf, .t64 and .adf as supported files (core not tested)
```
**RGB-Pi OS (Beta - Update) 2019 07 14**
```
- Changed VSync FIX from 3 to 4 (grid readjustment required)
- Fixed some Arcade games image distortion (i.e gigawing)
- Fixed DOSBox launcher and verbosity output
- Fixed Wifi firewall rules
- Fixed Remaps folder permissions
- Added help on rotating image to the left
- Increased keyrepeat response time
- Fixed Sega CD (JP) games not loading
```
**RGB-Pi OS (Beta - Image) 2019 06 30**
```
- Added game statistics information in Info screen
- Added Extras menu for special retroarch options
- Added new menu music modes and songs
- Added configuration upload option
- Added 2 speeds key repeat functionality in game lists
- Added 3s auto restore resolution on FIX option
- Added new scraper custom mode
- Added option to skip all analog buttons when mapping
- Added new gamepad sorting engine to match retroarch system
- Added quick skip to letter search functionality
- Added especial rom badges for pirate, unlicense, prototype, translated, hack, etc.
- Added new info popup in quick load functionality
- Added NAS/NFS option for reading games remotely
- Added new rules for some special gamepads
- Added country flag badges on gamelist
- Added new original/forced arcade selector
- Added filter to the game scanner to avoid displaying Linux/Mac hidden files
- Added improved file extension filter to DOSBox and ScummVM
- Added new automatic position adjustment of video intro
- Changed to new Bluetooth gamepad module. Now works properly and faster.
- Changed to new game search engine. Now is 2x faster
- Changed all UI image assets
- Changed UI hdmi_timings
- Changed/Fixed some UI texts
- Changed Netplay QR image by simple text URL
- Changed image Zoom algorithm
- Changed FIX VSync from 5 to 3 (grid readjustment required)
- Changed some buttons for screen assistant to fit with JAMMA board
- Updated help
- Updated JAMMA image
- Removed back/cancel option from screen assistant
- Removed testing BIOS folder for public Release
- Fixed lang selector on first boot for JAMMA board
- Fixed resolution change issue in the middle of video intro
- Fixed file permissions in Arcade folder
- Fixed sceen assistant not using the correct timing when FIX is active
- Fixed favorites not saving changes sometimes
- Fixed grid image proportions
- Fixed black screen happening sometimes in no-games screen
- Fixed analog sticks in retroarch mappings

python/mount_USB.py
#!/usr/bin/python
# coding: utf-8

# IMPORTS
import time
import pygame
import sys
import os
from pygame.locals import *

# INITS
pygame.init()
pygame.mouse.set_visible(0)

# VARIABLES
x_screen = 450
y_screen = 270

opt = 0
running = 1
x = 0
y = 0
y_slide = 0


# SET SCREEN
fullscreen = pygame.display.set_mode((x_screen,y_screen), RESIZABLE)
fullscreen.fill((165,165,255))	

# FONT
myfont = pygame.font.SysFont("Helvetica", 18)

# SHOW BACKGROUND 		
pygame.draw.rect(fullscreen, (66,66,231), (20,20,x_screen-40,y_screen-40), 0)

#title and credits
title = myfont.render("USB Storage mounted...", 1, (255,255,255))
inst = myfont.render("Create /roms/ folder on USB", 1, (255,255,255))
fullscreen.blit(title, (110, 100))
fullscreen.blit(inst, (110, 130))
pygame.display.flip()
time.sleep(5)


os.system('mount /media/usb0/roms /recalbox/share/roms')
os.system('/bin/bash /etc/init.d/S31emulationstation stop')
os.system('/bin/bash /etc/init.d/S31emulationstation start')
		
pygame.quit()
sys.exit()
	
	





python/Config.txt Editor.py
import pygame, sys, os, math, time, random
from pygame.locals import*

## CONSTANTS, yo ## 

DISPLAYWIDTH  = 400
DISPLAYHEIGHT = 220
FPS          = 50
TEXTHEIGHT   = 13
STARTX       = 0
STARTY       = 0
LEFT = 'left'
RIGHT = 'right'
UP = 'up'
DOWN = 'down'

## COLORS ##

#            R    G    B
GRAY     = (100, 100, 100)
NAVYBLUE = ( 60,  60, 100)
WHITE    = (255, 255, 255)
RED      = (255,   0,   0)
GREEN    = (  0, 255,   0)
DARKGREEN= (  0, 125,   0)
BLUE     = ( 66,  66, 231)
YELLOW   = (255, 255,   0)
ORANGE   = (255, 128,   0)
PURPLE   = (255,   0, 255)
CYAN     = (  0, 255, 255)
BLACK    = (  0,   0,   0)
COMBLUE  = (233, 232, 255)
BLUESWEET= (165, 165, 255)
ALPHA    = (  0,   0,   0,   0)

BGCOLOR = BLACK
TEXTCOLOR = WHITE

def main():
    global FPSCLOCK
    global loopCounter
    global loopList
    global directoryString
    pygame.init()
    pygame.joystick.init()
    FPSCLOCK = pygame.time.Clock()
    
    windowWidth  = 400
    windowHeight = 220 
    lineNumber   = 0
    newChar      = ''
    typeChar     = False
    textString   = " "
    mainList     = []
    deleteKey    = False
    returnKey    = False
    insertPoint  = 0
    camerax      = 0
    cameray      = 0
    mouseClicked = False
    mouseX       = 0
    mouseY       = 0
    loopList = ['-','0','1','2','3','4','5','6','7','8','9']
    loopCounter = [0]
    directoryString = ['/boot/config.txt']
    
    pygame.mouse.set_visible(0)
    displaySurf = pygame.display.set_mode((windowWidth, windowHeight), RESIZABLE)    
    displaySurf.fill(BGCOLOR)
    displaySurf.convert()
    pygame.display.update()
    
    pygame.display.set_caption('Notepad')
    mainFont = pygame.font.SysFont('Helvetica', TEXTHEIGHT)
    
    cursorRect = getCursorRect(STARTX, STARTY + (TEXTHEIGHT + (TEXTHEIGHT/4)), mainFont, camerax, cameray)
    
    mainList = LoadFile(mainList, windowWidth, windowHeight, displaySurf, mainFont)

    # 1ST JOYSTICK INIT
    joystick_one = pygame.joystick.Joystick(0)
    joystick_one.init()

## The main game loop detects user input, displays the text on the screen,
## displays the cursor on the screen, and adjusts the camera view if
## necessary.
    
    while True:
        
        camerax, cameray = adjustCamera(mainList, lineNumber, insertPoint, cursorRect, mainFont, camerax, cameray, windowWidth, windowHeight)

        newChar, typeChar, deleteKey, returnKey, directionKey, windowWidth, windowHeight, mouseX, mouseY, mouseClicked= getInput(windowWidth, windowHeight)
	

        if newChar == "escape":

	    os.system('mount -o remount,rw /boot')

            mainList = SaveFile(mainList, windowWidth, windowHeight, displaySurf, mainFont)

            newChar = False
            insertPoint = 0
            lineNumber = 0

	    os.system('mount -o remount,ro /boot')

	    pygame.quit()
            sys.exit()
		        
        mainList, lineNumber, insertPoint, cursorRect = displayText(mainFont, newChar, typeChar, mainList, deleteKey, returnKey, lineNumber, insertPoint, directionKey, camerax, cameray, cursorRect, windowWidth, windowHeight, displaySurf, mouseClicked, mouseX, mouseY)


        pygame.display.update()
        FPSCLOCK.tick(FPS)


## Interprets user input and changes mainList, lineNumber, insertPoint
## and cursorRect accordingly.  There is a function called blitAll()
## which blits all strings to the main surface.

def displayText(mainFont, newChar, typeChar, mainList, deleteKey, returnKey, lineNumber, insertPoint, directionKey, camerax, cameray, cursorRect, windowWidth, windowHeight, displaySurf, mouseClicked, mouseX, mouseY):
    if returnKey:
        firstString = getStringAtInsertPoint(mainList, lineNumber, insertPoint)
        secondString = getStringAfterInsertPoint(mainList, lineNumber, insertPoint)
        mainList[lineNumber] = firstString
        mainList.insert(lineNumber+1, secondString)
        lineNumber +=1
        returnKey = False
        insertPoint = 0
        cursorRect.x = STARTX
        stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
        cursorRect.y = stringRect.top
        
    elif mouseClicked:
        insertPoint, lineNumber, cursorRect = setCursorToClick(mainList, cursorRect, mainFont, camerax, cameray, mouseX, mouseY)

    elif directionKey:
        if directionKey == LEFT:
            if lineNumber == 0:
                if insertPoint > 0:
                    insertPoint -= 1
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = STARTY
                    
            elif lineNumber > 0:
                if insertPoint == 0:
                    lineNumber -= 1
                    insertPoint = len(mainList[lineNumber])
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = stringRect.top
                    
                elif insertPoint > 0:
                    insertPoint -= 1
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)

                    if insertPoint == 0:
                        cursorRect.x = STARTX
                        cursorRect.y = stringRect.top
                    else:
                        cursorRect.x = stringRect.right
                        cursorRect.y = stringRect.top
                    
        elif directionKey == RIGHT:
            if insertPoint < len(mainList[lineNumber]):
                insertPoint += 1
                stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                cursorRect.x = stringRect.right
                cursorRect.y = stringRect.top

            elif insertPoint >= len(mainList[lineNumber]):
                if len(mainList) > (lineNumber + 1):
                    lineNumber += 1
                    insertPoint = 0
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = stringRect.top
                    
        elif directionKey == UP:
            if lineNumber > 0:
                if insertPoint == 0:
                    lineNumber -= 1
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = STARTX
                    cursorRect.y = stringRect.top
                    
                elif insertPoint > len(mainList[lineNumber - 1]):
                    lineNumber -= 1
                    insertPoint = len(mainList[lineNumber])
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = stringRect.top
                      
                elif insertPoint <= len(mainList[lineNumber -1]):
                    lineNumber -= 1
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = stringRect.top
                    
        elif directionKey == DOWN:
            if lineNumber + 1 < len(mainList):
                if insertPoint == 0:
		    lineNumber += 1
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = STARTX
                    cursorRect.y = stringRect.top
                    
                elif insertPoint > len(mainList[lineNumber + 1]):
                    lineNumber +=1
                    insertPoint = len(mainList[lineNumber])
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = stringRect.top
                elif insertPoint <= len(mainList[lineNumber +1]):
                    lineNumber += 1
                    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
                    cursorRect.x = stringRect.right
                    cursorRect.y = stringRect.top
                    
    elif typeChar:
        string = mainList[lineNumber]
        stringList = list(string)
        stringList.insert(insertPoint, newChar)
        newString = ''.join(stringList)
        mainList[lineNumber] = newString
        
        typeChar = False

        if len(newString) > len(string) and newChar != '    ':   ## Prevents alteration keys like shifts from affecting the insertPoint ##
            insertPoint += 1
            stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
            cursorRect.x = stringRect.right
            cursorRect.y = stringRect.top
            
        elif newChar == '    ':
            insertPoint += 4
            stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
            cursorRect.x = stringRect.right
            cursorRect.y = stringRect.top

    elif deleteKey:
        
        if insertPoint > 0:
            firstString = getStringAtInsertPoint(mainList, lineNumber, insertPoint)
            secondString = getStringAfterInsertPoint(mainList, lineNumber, insertPoint)
            stringList = list(firstString)
            del stringList[insertPoint-1]
            string = ''.join(stringList)
            string += secondString
            mainList[lineNumber] = string

            deleteKey = False
            insertPoint -= 1
            stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
            cursorRect.x = stringRect.right
            cursorRect.y = stringRect.top
                                
        elif insertPoint <= 0:
            if lineNumber > 0:
                string = getStringAfterInsertPoint(mainList, lineNumber, insertPoint)
                del mainList[lineNumber]
                lineNumber -= 1
                mainList[lineNumber] += string
                
                deleteKey = False
                insertPoint = len(mainList[lineNumber]) - len(string)
                stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)                
                cursorRect.x = stringRect.right
                cursorRect.y = stringRect.top

    else:
        stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)

        if insertPoint == 0:
            cursorRect.x = STARTX
        elif insertPoint > 0:
            cursorRect.x = stringRect.right
        if lineNumber == 0:
            cursorRect.y = STARTY
        elif lineNumber > 0:
            cursorRect.y = stringRect.top
        else:
            cursorRect.x = stringRect.right

    if cursorRect.left >= STARTX:
        if cursorRect.right <= windowWidth:
            if cursorRect.top >= STARTY:
                if cursorRect.bottom <= (windowHeight - STARTY):
                    blitAll(mainList, mainFont, camerax, cameray, cursorRect, displaySurf)
    
    return mainList, lineNumber, insertPoint, cursorRect

##################################################################
## Blits all the strings in mainList to the main surface object 
##################################################################
##################################################################


def blitAll(mainList, mainFont, camerax, cameray, cursorRect, displaySurf):
    displaySurf.fill(BGCOLOR)
    i = 0
    
    for string in mainList:  ##blitting all the strings in the mainList by iterating through them
        stringRender = mainFont.render(string, True, TEXTCOLOR, BGCOLOR)
        stringRect = stringRender.get_rect()
        stringRect.x = STARTX - camerax
        stringRect.y = STARTY + (i * (TEXTHEIGHT + (TEXTHEIGHT/4))) - cameray
        displaySurf.blit(stringRender, stringRect)
        i += 1

    drawCursor(mainFont, cursorRect, displaySurf)


def adjustCamera(mainList, lineNumber, insertPoint, cursorRect, mainFont, camerax, cameray, windowWidth, windowHeight):

    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)
    
    if (stringRect.right + cursorRect.width) > windowWidth:
        camerax += (stringRect.right + cursorRect.width) - windowWidth  
    elif cursorRect.left < STARTX:
        camerax -= (-1)*(cursorRect.left)

    if stringRect.bottom > windowHeight:
        cameray += stringRect.bottom - windowHeight
    elif stringRect.top < 0:
        cameray -= (-1)*(stringRect.top)
               
    if insertPoint == 0:
        camerax = 0
    if lineNumber == 0:
        cameray = 0

    return camerax, cameray
    

def drawCursor(mainFont, cursorRect, displaySurf):
    cursor = mainFont.render("<<<", True, BLACK, WHITE)
    displaySurf.blit(cursor, cursorRect)
    



def getInput(windowWidth, windowHeight):
    newChar = False
    typeChar = False
    deleteKey = False
    returnKey = False
    directionKey = False
    mouseX = 0
    mouseY = 0
    mouseClicked = False
        
    for event in pygame.event.get():
	
        if event.type == QUIT:
            pygame.quit()
            sys.exit()

        elif event.type == JOYBUTTONDOWN:
		if event.button == 0:
			deleteKey = True
		elif event.button == 1:
			deleteKey = True
		elif event.button == 2:
			deleteKey = True
		elif event.button == 3:
			newChar = ' '
			typeChar = True
		elif event.button == 8:
			newChar = 'escape'
			

        elif event.type == JOYBUTTONUP:
		if event.button == 1:
			if loopCounter[0] < 10:
				loopCounter[0]+=1
			else:
				loopCounter[0]=0
			newChar = loopList[loopCounter[0]]
                	typeChar = True
		elif event.button == 2:
			if loopCounter[0] == 0:
				loopCounter[0]=10
			else:
				loopCounter[0]-=1
			newChar = loopList[loopCounter[0]]
                	typeChar = True
		


	elif event.type == JOYAXISMOTION:
		if event.axis == 1 and event.value > 0:
			directionKey = DOWN
		elif event.axis == 1 and event.value < 0:
			directionKey = UP
		elif event.axis == 0 and event.value > 0:
			directionKey = RIGHT
		elif event.axis == 0 and event.value < 0:
			directionKey = LEFT

            
        elif event.type == KEYDOWN:
            if event.key == K_BACKSPACE:
                deleteKey = True
            elif event.key == K_ESCAPE:
                newChar = 'escape'
            elif event.key == K_RETURN:
                returnKey = True
            elif event.key == K_TAB:
                newChar = '    '
                typeChar = True
            elif event.key == K_LEFT:
                directionKey = LEFT
            elif event.key == K_RIGHT:
                directionKey = RIGHT
            elif event.key == K_UP:
                directionKey = UP
            elif event.key == K_DOWN:
		directionKey = DOWN

	    else:
                newChar = event.unicode
                typeChar = True
			

        elif event.type == VIDEORESIZE:
            displaySurf = pygame.display.set_mode(event.dict['size'], RESIZABLE)
            windowWidth = event.dict['w']
            windowHeight = event.dict['h']
            displaySurf.fill(BLUE)
            displaySurf.convert()
            pygame.display.update()
            
           
            
        elif event.type == MOUSEBUTTONDOWN:
            mouseX, mouseY = event.pos
            #mouseClicked = True
	    directionKey = DOWN

    return newChar, typeChar, deleteKey, returnKey, directionKey, windowWidth, windowHeight, mouseX, mouseY, mouseClicked


## These functions involve the string the cursor happens to be on.
## By using the lineNumber, the program knows which string to
## manipulate.  lineNumber = 0 is the first line, and so on.
## The cursor's left position is always locked to the right of whatever
## stringRect it is next to.

def getStringRect(string, lineNumber, camerax, cameray):
    stringRect = string.get_rect()
    stringRect.x = STARTX - camerax
    stringRect.y = STARTY + (lineNumber * (TEXTHEIGHT + (TEXTHEIGHT/4))) - cameray

    return stringRect

def getStringAtInsertPoint(mainList, lineNumber, insertPoint):
    string = mainList[lineNumber]
    stringList = list(string)
    newStringList = stringList[0:insertPoint]
    newString = ''.join(newStringList)

    return newString

def getStringAfterInsertPoint(mainList, lineNumber, insertPoint):
    string = mainList[lineNumber]
    stringList = list(string)
    newStringList = stringList[insertPoint:]
    newString = ''.join(newStringList)

    return newString


def getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray):
    string = getStringAtInsertPoint(mainList, lineNumber, insertPoint)
    stringRender = mainFont.render(string, True, TEXTCOLOR, BGCOLOR)
    stringRect = getStringRect(stringRender, lineNumber, camerax, cameray)

    return stringRect


## Miscelaneous Functions.  getCursorRect is used to produce the
## cursors's rect object.


def getCursorRect(cursorX, cursorY, mainFont, camerax, cameray):
    cursor = mainFont.render('L', True, RED)
    cursorRect = cursor.get_rect()
    cursorRect.x = cursorX - camerax
    cursorRect.y = cursorY - cameray

    return cursorRect


## These three functions, setCursorToClick(), getLineNumberOfClick(), and
## get insertPointAtMouseX() allow the user to set the cursor location
## by clicking the mouse at the spot they want to go.


def setCursorToClick(mainList, cursorRect, mainFont, camerax, cameray, mouseX, mouseY):
    lineNumber = getLineNumberOfClick(mouseY, cameray, mainList)
    insertPoint = getInsertPointAtMouseX(mouseX, mouseY, lineNumber, mainList, mainFont, camerax, cameray)
    stringRect = getStringRectAtInsertPoint(mainList, lineNumber, insertPoint, mainFont, camerax, cameray)

    if insertPoint == 0:
        cursorRect.x = STARTX
    elif insertPoint > 0:
        cursorRect.x = stringRect.right
        
    cursorRect.y = stringRect.top

    return insertPoint, lineNumber, cursorRect


def getLineNumberOfClick(mouseY, cameray, mainList):
    clickLineNumber = (mouseY + cameray) / float(TEXTHEIGHT+ (TEXTHEIGHT/4))
    if clickLineNumber > len(mainList):
        lineNumber = (len(mainList)) - 1
    elif clickLineNumber <= len(mainList):
        floorLineNumber = math.floor(clickLineNumber)
        lineNumber = int(floorLineNumber)

    return lineNumber


def getInsertPointAtMouseX(mouseX, mouseY, lineNumber, mainList, mainFont, camerax, cameray):
    string = mainList[lineNumber]
    newInsertPoint = 0

    if (mouseY + cameray) > ((lineNumber + 1) * (TEXTHEIGHT + TEXTHEIGHT/4)):
        insertPoint = len(mainList[lineNumber])
        return insertPoint
    
    for insertPoint in string:
        stringRect = getStringRectAtInsertPoint(mainList, lineNumber, newInsertPoint, mainFont, camerax, cameray)
        if mouseX >= stringRect.left:
            if mouseX < stringRect.right:
                if newInsertPoint > 0:
                    return newInsertPoint - 1

        newInsertPoint += 1

    else:
        return newInsertPoint


def LoadFile(mainList, windowWidth, windowHeight, displaySurf, mainFont):
    

    loadDirectory = False

    loadDirectory, directoryString[0] = directoryInput(displaySurf, directoryString[0], mainFont, loadDirectory)
                
    newMainList = loadFromDisk(loadDirectory)
    return newMainList

    return mainList


def SaveFile(mainList, windowWidth, windowHeight, displaySurf, mainFont):
    

    saveDirectory = False
            
    saveDirectory, directoryString[0] = directoryInput(displaySurf, directoryString[0], mainFont, saveDirectory)
                
    saveToDisk(mainList, saveDirectory)

    
    return mainList
    
    

def directoryInput(displaySurf, directoryString, mainFont, saveOrLoadDirectory):
    directoryList = ''.join(directoryString)

    saveOrLoadDirectory = directoryString

    return saveOrLoadDirectory, directoryString


def saveToDisk(mainList, saveDirectory):
    saveFile = open(saveDirectory, 'w')

    for string in mainList:
        saveFile.write(string + '\n')

    saveFile.close()


def loadFromDisk(loadDirectory):
    mainList = []
    saveFile = open(loadDirectory, 'r')

    for line in saveFile:
        mainList.append(line)

    saveFile.close()

    i = 0
    while i < len(mainList):
        stringList = list(mainList[i])
        stringList.pop()
        mainList[i] = ''.join(stringList)
        i += 1

    return mainList



    
def getStringRenderAndRect(string, mainFont):
    stringRender = mainFont.render(string, True, TEXTCOLOR, BLUE)
    stringRect = stringRender.get_rect()

    return stringRender, stringRect


if __name__ == '__main__':
    main()


